Generalize gesvdjBatched to run whith full_matrices==false

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88502

As brought up in https://github.com/pytorch/pytorch/issues/86234#issuecomment-1268296036, our heuristic for which SVD backend to choose was not great in some cases.
The case in which there could be some improvements is when we have a
large batch of very small non-square matrices.

This PR, adapts the calling code to gesvdj by creating two temporary
square buffers to allow to call gesvdjBatched, and then copies back the
result into the output buffers.

We then modify the heuristic that chooses between gesvdj and
gesvdjBatched.

cc @VitalyFedyunin @ngimel @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @JulianKnodt

Fixes https://github.com/pytorch/pytorch/issues/86234

lezcano committed:
Generalize gesvdjBatched to run whith full_matrices==false

As brought up in https://github.com/pytorch/pytorch/issues/86234#issuecomment-1268296036, our heuristic for which SVD backend to choose was not great in some cases.
The case in which there could be some improvements is when we have a
large batch of very small non-square matrices.

This PR, adapts the calling code to gesvdj by creating two temporary
square buffers to allow to call gesvdjBatched, and then copies back the
result into the output buffers.

We then modify the heuristic that chooses between gesvdj and
gesvdjBatched.

cc @JulianKnodt

Fixes https://github.com/pytorch/pytorch/issues/86234

[ghstack-poisoned]

lezcano commented:
The benchmarks show that the batched version is always faster:
```
[--------------------- SVD ----------------------]
                          |  batched  |  iterative
32 threads: --------------------------------------
      shape (2, 4)       |     87    |     190
      shape (2, 8)       |     87    |     190
      shape (2, 16)      |     90    |     191
      shape (2, 32)      |     94    |     200
      shape (4, 2)       |     89    |     190
      shape (4, 4)       |    140    |     254
      shape (4, 8)       |    130    |     292
      shape (4, 16)      |    132    |     291
      shape (4, 32)      |    140    |     301
      shape (8, 2)       |     80    |     190
      shape (8, 4)       |    150    |     289
      shape (8, 8)       |    172    |     381
      shape (8, 16)      |    177    |     375
      shape (8, 32)      |    190    |     375
      shape (16, 2)      |     85    |     190
      shape (16, 4)      |    130    |     289
      shape (16, 8)      |    173    |     416
      shape (16, 16)     |    246    |     475
      shape (16, 32)     |    286    |     485
      shape (32, 2)      |     96    |     190
      shape (32, 4)      |    130    |     290
      shape (32, 8)      |    209    |     417
      shape (32, 16)     |    282    |     482
      shape (32, 32)     |    296    |     585
      shape (1, 2, 2)    |     77    |     160
      shape (1, 2, 4)    |     93    |     190
      shape (1, 2, 8)    |     93    |     190
      shape (1, 2, 16)   |     92    |     190
      shape (1, 2, 32)   |     95    |     190
      shape (1, 4, 2)    |     90    |     190
      shape (1, 4, 4)    |    120    |     300
      shape (1, 4, 8)    |    130    |     292
      shape (1, 4, 16)   |    140    |     293
      shape (1, 4, 32)   |    160    |     293
      shape (1, 8, 2)    |     91    |     190
      shape (1, 8, 4)    |    130    |     290
      shape (1, 8, 8)    |    170    |     382
      shape (1, 8, 16)   |    200    |     376
      shape (1, 8, 32)   |    190    |     376
      shape (1, 16, 2)   |     90    |     190
      shape (1, 16, 4)   |    135    |     291
      shape (1, 16, 8)   |    190    |     374
      shape (1, 16, 16)  |    244    |     484
      shape (1, 16, 32)  |    278    |     486
      shape (1, 32, 2)   |     96    |     190
      shape (1, 32, 4)   |    140    |     290
      shape (1, 32, 8)   |    190    |     374
      shape (1, 32, 16)  |    283    |     475
      shape (1, 32, 32)  |    329    |     594
      shape (2, 2, 2)     |    120    |      311
      shape (2, 2, 4)     |     81    |      319
      shape (2, 2, 8)     |     81    |      319
      shape (2, 2, 16)    |     81    |      320
      shape (2, 2, 32)    |     92    |      321
      shape (2, 4, 2)     |     79    |      359
      shape (2, 4, 4)     |    132    |      497
      shape (2, 4, 8)     |    120    |      564
      shape (2, 4, 16)    |    130    |      516
      shape (2, 4, 32)    |    138    |      518
      shape (2, 8, 2)     |     79    |      314
      shape (2, 8, 4)     |    130    |      512
      shape (2, 8, 8)     |    174    |      649
      shape (2, 8, 16)    |    190    |      680
      shape (2, 8, 32)    |    205    |      680
      shape (2, 16, 2)    |     79    |      315
      shape (2, 16, 4)    |    140    |      512
      shape (2, 16, 8)    |    170    |      680
      shape (2, 16, 16)   |    249    |      880
      shape (2, 16, 32)   |    279    |      900
      shape (2, 32, 2)    |     87    |      316
      shape (2, 32, 4)    |    136    |      513
      shape (2, 32, 8)    |    186    |      690
      shape (2, 32, 16)   |    277    |      890
      shape (2, 32, 32)   |    323    |     1100
      shape (4, 2, 2)     |     92    |      570
      shape (4, 2, 4)     |    100    |      632
      shape (4, 2, 8)     |     85    |      588
      shape (4, 2, 16)    |    100    |      587
      shape (4, 2, 32)    |     91    |      588
      shape (4, 4, 2)     |     83    |      576
      shape (4, 4, 4)     |    130    |      840
      shape (4, 4, 8)     |    130    |      979
      shape (4, 4, 16)    |    132    |      980
      shape (4, 4, 32)    |    136    |      977
      shape (4, 8, 2)     |     80    |      576
      shape (4, 8, 4)     |    130    |      970
      shape (4, 8, 8)     |    170    |     1350
      shape (4, 8, 16)    |    170    |     1400
      shape (4, 8, 32)    |    188    |     1370
      shape (4, 16, 2)    |     80    |      621
      shape (4, 16, 4)    |    130    |      970
      shape (4, 16, 8)    |    190    |     1310
      shape (4, 16, 16)   |    250    |     1800
      shape (4, 16, 32)   |    296    |     1800
      shape (4, 32, 2)    |    100    |      578
      shape (4, 32, 4)    |    137    |      970
      shape (4, 32, 8)    |    203    |     1400
      shape (4, 32, 16)   |    278    |     1800
      shape (4, 32, 32)   |    323    |     2221
      shape (8, 2, 2)     |     93    |     1090
      shape (8, 2, 4)     |     86    |     1300
      shape (8, 2, 8)     |     99    |     1100
      shape (8, 2, 16)    |     83    |     1200
      shape (8, 2, 32)    |     90    |     1100
      shape (8, 4, 2)     |     96    |     1190
      shape (8, 4, 4)     |    130    |     1700
      shape (8, 4, 8)     |    130    |     1942
      shape (8, 4, 16)    |    130    |     1890
      shape (8, 4, 32)    |    138    |     1900
      shape (8, 8, 2)     |     79    |     1096
      shape (8, 8, 4)     |    140    |     1800
      shape (8, 8, 8)     |    174    |     2518
      shape (8, 8, 16)    |    190    |     2557
      shape (8, 8, 32)    |    206    |     2598
      shape (8, 16, 2)    |     80    |     1140
      shape (8, 16, 4)    |    130    |     1869
      shape (8, 16, 8)    |    190    |     2537
      shape (8, 16, 16)   |    243    |     3416
      shape (8, 16, 32)   |    303    |     3590
      shape (8, 32, 2)    |     87    |     1100
      shape (8, 32, 4)    |    140    |     1830
      shape (8, 32, 8)    |    203    |     2580
      shape (8, 32, 16)   |    300    |     3469
      shape (8, 32, 32)   |    322    |     4340
      shape (16, 2, 2)    |     92    |     2047
      shape (16, 2, 4)    |     84    |     2275
      shape (16, 2, 8)    |     82    |     2234
      shape (16, 2, 16)   |     85    |     2182
      shape (16, 2, 32)   |    100    |     2253
      shape (16, 4, 2)    |     82    |     2276
      shape (16, 4, 4)    |    130    |     3446
      shape (16, 4, 8)    |    129    |     3816
      shape (16, 4, 16)   |    130    |     3772
      shape (16, 4, 32)   |    160    |     3723
      shape (16, 8, 2)    |     96    |     2231
      shape (16, 8, 4)    |    128    |     3691
      shape (16, 8, 8)    |    172    |     5208
      shape (16, 8, 16)   |    190    |     5320
      shape (16, 8, 32)   |    206    |     5135
      shape (16, 16, 2)   |     99    |     2138
      shape (16, 16, 4)   |    130    |     3733
      shape (16, 16, 8)   |    190    |     5095
      shape (16, 16, 16)  |    249    |     6779
      shape (16, 16, 32)  |    303    |     6920
      shape (16, 32, 2)   |    100    |     2164
      shape (16, 32, 4)   |    153    |     3640
      shape (16, 32, 8)   |    203    |     5228
      shape (16, 32, 16)  |    276    |     6900
      shape (16, 32, 32)  |    326    |     8680
      shape (32, 2, 2)    |     92    |     3912
      shape (32, 2, 4)    |    110    |     4397
      shape (32, 2, 8)    |     99    |     4517
      shape (32, 2, 16)   |    106    |     4395
      shape (32, 2, 32)   |    111    |     4370
      shape (32, 4, 2)    |     96    |     4373
      shape (32, 4, 4)    |    160    |     7000
      shape (32, 4, 8)    |    160    |     7400
      shape (32, 4, 16)   |    160    |     7400
      shape (32, 4, 32)   |    190    |     7500
      shape (32, 8, 2)    |    100    |     4396
      shape (32, 8, 4)    |    160    |     7410
      shape (32, 8, 8)    |    216    |    10300
      shape (32, 8, 16)   |    238    |    10000
      shape (32, 8, 32)   |    246    |    10000
      shape (32, 16, 2)   |    110    |     4281
      shape (32, 16, 4)   |    161    |     7400
      shape (32, 16, 8)   |    236    |    10000
      shape (32, 16, 16)  |    286    |    13500
      shape (32, 16, 32)  |    377    |    13800
      shape (32, 32, 2)   |    110    |     4364
      shape (32, 32, 4)   |    170    |     7300
      shape (32, 32, 8)   |    244    |    10000
      shape (32, 32, 16)  |    350    |    13700
      shape (32, 32, 32)  |    431    |    17000
      shape (64, 2, 2)    |    110    |     7800
      shape (64, 2, 4)    |    120    |     8750
      shape (64, 2, 8)    |    120    |     8800
      shape (64, 2, 16)   |    130    |     8700
      shape (64, 2, 32)   |    130    |     8800
      shape (64, 4, 2)    |    115    |     8570
      shape (64, 4, 4)    |    222    |    14000
      shape (64, 4, 8)    |    214    |    15000
      shape (64, 4, 16)   |    222    |    14800
      shape (64, 4, 32)   |    239    |    15000
      shape (64, 8, 2)    |    120    |     8600
      shape (64, 8, 4)    |    211    |    14800
      shape (64, 8, 8)    |    315    |    20560
      shape (64, 8, 16)   |    328    |    21060
      shape (64, 8, 32)   |    345    |    20980
      shape (64, 16, 2)   |    140    |     8500
      shape (64, 16, 4)   |    215    |    14500
      shape (64, 16, 8)   |    313    |    20590
      shape (64, 16, 16)  |    463    |    26730
      shape (64, 16, 32)  |    566    |    27350
      shape (64, 32, 2)   |    130    |     8500
      shape (64, 32, 4)   |    253    |    14400
      shape (64, 32, 8)   |    350    |    20450
      shape (64, 32, 16)  |    563    |    27130
      shape (64, 32, 32)  |    657    |    34260

Times are in microseconds (us).
```

lezcano committed:
Update on "Generalize gesvdjBatched to run whith full_matrices==false"


As brought up in https://github.com/pytorch/pytorch/issues/86234#issuecomment-1268296036, our heuristic for which SVD backend to choose was not great in some cases.
The case in which there could be some improvements is when we have a
large batch of very small non-square matrices.

This PR, adapts the calling code to gesvdj by creating two temporary
square buffers to allow to call gesvdjBatched, and then copies back the
result into the output buffers.

We then modify the heuristic that chooses between gesvdj and
gesvdjBatched.

cc VitalyFedyunin ngimel jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano JulianKnodt

Fixes https://github.com/pytorch/pytorch/issues/86234

[ghstack-poisoned]

IvanYashchuk reviewed (approved):
Great!

lezcano commented:
@pytorchbot merge

nikitaved commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -695,10 +693,42 @@ inline static void apply_svd_cusolver_gesvdjBatched(const Tensor& A, const Tenso
+      U_ = U.new_empty(sizes).mT();
+    } else if (m < n) {
+      // Size of V with full_matrices == True
+      sizes.end()[-2] = n;
+      sizes.end()[-1] = n;
```
nit: `sizes.end()[-1] = n` is redundant.

nikitaved commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -695,10 +693,42 @@ inline static void apply_svd_cusolver_gesvdjBatched(const Tensor& A, const Tenso
+  auto m = A.size(-2);
+  auto n = A.size(-1);
+  auto k = std::min(m, n);
+  // The kernel assumes full_matrices == true
+  // If full_matrices == false, we create auxiliary tensors of the right size and copy the results back
```
nit:  This is true only if `m != n`. Let's probably add that to reduce confusion?

nikitaved commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -695,10 +693,42 @@ inline static void apply_svd_cusolver_gesvdjBatched(const Tensor& A, const Tenso
+    apply_svd_cusolver_gesvdjBatched<scalar_t>(A, U_, S, V_, infos, compute_uv);
   });
+
+  // Copy the result back
+  if (compute_uv && !full_matrices) {
+    if (m > n) {
+      U.copy_(U_.narrow(-1, 0, k));
+    } else if (m < n) {
+      V.copy_(V_.narrow(-1, 0, k));
+    }
+  }
```
nit: same as above.

lezcano replied:
changed the conditions to avoid code repetition.

nikitaved reviewed (approved):
Looks awesome! Thank you, @lezcano ! I have just left a couple of minor nits.

mruberry reviewed (approved):
Cool! Stamped!

mruberry commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -832,21 +862,23 @@ void svd_cusolver(const Tensor& A,
-    if (m <= 32 && n <= 32 && batch_size > 1 && (full_matrices || m == n)) {
-      svd_cusolver_gesvdjBatched(cloneBatchedColumnMajor(A), U, S, V, info, compute_uv);
+    // See the benchmarks in
+    // https://github.com/pytorch/pytorch/pull/88502#issuecomment-1303860789
```
Nice comment to explain the heuristics

lezcano replied:
done

xwang233 replied:
It's interesting to know that the batched 1x32x32 is faster than the non-batched version. 🤔

lezcano replied:
I reckon it's because it may be doing less? That is, perhaps the other function is a bit more precise? It'd be good to have a look internally.

lezcano committed:
Update on "Generalize gesvdjBatched to run whith full_matrices==false"


As brought up in https://github.com/pytorch/pytorch/issues/86234#issuecomment-1268296036, our heuristic for which SVD backend to choose was not great in some cases.
The case in which there could be some improvements is when we have a
large batch of very small non-square matrices.

This PR, adapts the calling code to gesvdj by creating two temporary
square buffers to allow to call gesvdjBatched, and then copies back the
result into the output buffers.

We then modify the heuristic that chooses between gesvdj and
gesvdjBatched.

cc VitalyFedyunin ngimel jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano JulianKnodt

Fixes https://github.com/pytorch/pytorch/issues/86234

[ghstack-poisoned]

lezcano committed:
Update on "Generalize gesvdjBatched to run whith full_matrices==false"


As brought up in https://github.com/pytorch/pytorch/issues/86234#issuecomment-1268296036, our heuristic for which SVD backend to choose was not great in some cases.
The case in which there could be some improvements is when we have a
large batch of very small non-square matrices.

This PR, adapts the calling code to gesvdj by creating two temporary
square buffers to allow to call gesvdjBatched, and then copies back the
result into the output buffers.

We then modify the heuristic that chooses between gesvdj and
gesvdjBatched.

cc VitalyFedyunin ngimel jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano JulianKnodt

Fixes https://github.com/pytorch/pytorch/issues/86234

[ghstack-poisoned]

lezcano commented:
@IvanYashchuk I deleted a test that was testing the behaviour of the `svd` function when given `NaN`s. This behaviour is not defined in the docs, and it's not consistent between backends or drivers, so I don't think we should test it. For example, this new backend does not raise an error when an input with nans is provided, it simply returns `S = [nan, nan, ...]`.

mruberry commented on test/test_linalg.py:
```diff
@@ -2476,28 +2476,6 @@ def test_svd_memory_allocation(self, device, dtype):
-    @skipCUDAIfNoCusolver
-    @skipCPUIfNoLapack
-    @dtypes(*floating_and_complex_types())
-    def test_svd_nan_error(self, device, dtype):
```
Is there something we can do (a documentation update or a comment in the code or OpInfo?) that will spare future developers adding this test again?

lezcano replied:
Merge this PR https://github.com/pytorch/pytorch/pull/88313 :P

lezcano committed:
Update on "Generalize gesvdjBatched to run whith full_matrices==false"


As brought up in https://github.com/pytorch/pytorch/issues/86234#issuecomment-1268296036, our heuristic for which SVD backend to choose was not great in some cases.
The case in which there could be some improvements is when we have a
large batch of very small non-square matrices.

This PR, adapts the calling code to gesvdj by creating two temporary
square buffers to allow to call gesvdjBatched, and then copies back the
result into the output buffers.

We then modify the heuristic that chooses between gesvdj and
gesvdjBatched.

cc VitalyFedyunin ngimel jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano JulianKnodt

Fixes https://github.com/pytorch/pytorch/issues/86234

[ghstack-poisoned]

zou3519 commented on test/functorch/test_vmap.py:
```diff
@@ -3236,8 +3236,6 @@ def test():
         xfail('broadcast_shapes', ''),  # test runner can't handle non-Tensor ops
         xfail('sparse.sampled_addmm'),  # sparse
         xfail('cross'),  # The default value of dim in op is *very* weird. No wonder it doesn't work
-        xfail('svd', device_type='cuda'),  # not unique, see test_linalg_svd for manual test
-        xfail('linalg.svd', device_type='cuda'),  # not unique, see test_linalg_svd for manual test
```
Does linalg.svd produce the same result for a matrix and a batch of matrices containing that matrix?

Previously the answer was no, hence the xfail. If with the new implementation this is guaranteed then removing the xfail is fine; if it's not guaranteed then it might be better to change this to a skip

zou3519 replied:
Let's make it a skip, someone might add a larger input size to OpInfo in the future and be confused when the test fails

lezcano replied:
Actually, nevermind, the answer to this question is now always yes, as it'll always use the same algorithm regardless of whether the matrix is a batch or not.

lezcano replied:
I'm removing the xfail and I won't tag it as a skip.

zou3519 replied:
SGTM

xwang233 reviewed (approved):
LGTM. Thanks for the work!

lezcano committed:
Update on "Generalize gesvdjBatched to run whith full_matrices==false"


As brought up in https://github.com/pytorch/pytorch/issues/86234#issuecomment-1268296036, our heuristic for which SVD backend to choose was not great in some cases.
The case in which there could be some improvements is when we have a
large batch of very small non-square matrices.

This PR, adapts the calling code to gesvdj by creating two temporary
square buffers to allow to call gesvdjBatched, and then copies back the
result into the output buffers.

We then modify the heuristic that chooses between gesvdj and
gesvdjBatched.

cc VitalyFedyunin ngimel jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano JulianKnodt

Fixes https://github.com/pytorch/pytorch/issues/86234

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge
